File: week51/wed/BOJ1520.java
package me.algo.week51.wed;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 * Created by Bomi on 2021/04/26.
 * 문제 출처 : https://www.acmicpc.net/problem/1520
 *
 * Time Complexity : O(N^2)
 * Used Algorithm : DP
 * Used Data structure : Array
 */

public class BOJ1520 {
    private static final int NOT_VISITED = -1;
    private static int height;
    private static int width;

    private static int[] dx = {-1, 1, 0, 0};
    private static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        height = Integer.parseInt(st.nextToken());
        width = Integer.parseInt(st.nextToken());

        int[][] map = new int[height][width];
        int[][] dp = new int[height][width];
        for (int i = 0; i < height; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < width; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                dp[i][j] = NOT_VISITED;
            }
        }

        System.out.println(dfs(0, 0, map, dp));
    }

    private static int dfs(int x, int y, int[][] map, int[][] dp) {
        if (x == (height - 1) && y == (width - 1)) return 1;
        if (dp[x][y] == NOT_VISITED) {
            dp[x][y] = 0;
            for (int i = 0; i < 4; i++) {
                int nextX = x + dx[i];
                int nextY = y + dy[i];

                if (nextX < 0 || nextX >= height || nextY < 0 || nextY >= width) continue;
                if (map[x][y] <= map[nextX][nextY]) continue;
                dp[x][y] += dfs(nextX, nextY, map, dp);
            }
        }
        return dp[x][y];
    }
}

File: week51/wed/BOJ1520.kt
package me.algo.week51.wed

import java.io.InputStreamReader

/**
 * Created by Bomi on 2021/04/26.
 * 문제 출처 : https://www.acmicpc.net/problem/1520
 *
 * Time Complexity : O(N^2)
 * Used Algorithm : DP
 * Used Data structure : Array
 */

private const val NOT_VISITED = -1

fun main() = with(InputStreamReader(System.`in`)) {
    val (height, width) = readLine()!!.split(" ").map { it.toInt() }
    val map = Array(height) { IntArray(width) }
    val dp = Array(height) { IntArray(width) }

    for (i in 0 until height) {
        readLine()!!.split(" ").forEachIndexed { j, it ->
            map[i][j] = it.toInt()
            dp[i][j] = NOT_VISITED
        }
    }
    println(dfs(0, 0, map, dp))
}

private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)

private fun dfs(x: Int, y: Int, map: Array<IntArray>, dp: Array<IntArray>): Int {
    if (x == (map.size - 1) && y == (map[0].size - 1)) return 1
    if (dp[x][y] == -1) {
        dp[x][y] = 0
        for (i in 0 until 4) {
            val nextX = x + dx[i]
            val nextY = y + dy[i]

            if (nextX < 0 || nextX >= map.size || nextY < 0 || nextY >= map[0].size) continue
            if (map[x][y] <= map[nextX][nextY]) continue
            dp[x][y] += dfs(nextX, nextY, map, dp)
        }
    }
    return dp[x][y]
}